junit5/src/main/java/cz/xtf/junit5/extensions/OpenShiftRecorderService.java

```diff
@@ -0,0 +1,527 @@
+package cz.xtf.junit5.extensions;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.platform.commons.support.AnnotationSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import cz.xtf.core.bm.BuildManagers;
+import cz.xtf.core.openshift.OpenShift;
+import cz.xtf.core.openshift.OpenShifts;
+import cz.xtf.junit5.annotations.OpenShiftRecorder;
+import cz.xtf.junit5.config.JUnitConfig;
+import cz.xtf.junit5.extensions.helpers.EventsFilterBuilder;
+import cz.xtf.junit5.extensions.helpers.ResourcesFilterBuilder;
+import cz.xtf.junit5.extensions.helpers.ResourcesPrinterHelper;
+import cz.xtf.junit5.extensions.helpers.ResourcesTimestampHelper;
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.api.model.Event;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import io.fabric8.openshift.api.model.Build;
+import io.fabric8.openshift.api.model.BuildConfig;
+import io.fabric8.openshift.api.model.DeploymentConfig;
+import io.fabric8.openshift.api.model.ImageStream;
+import io.fabric8.openshift.api.model.Route;
+
+/**
+ * Record OpenShift isolated state relative to a test.
+ * Provides an API and implementation for setting filtered resources and recording their state upon request.
+ *
+ * This service class can typically be used by classes that implement Junit extension lifecycle interfaces,
+ * e.g.: {@link org.junit.jupiter.api.extension.TestWatcher}, {@link org.junit.jupiter.api.extension.BeforeAllCallback}
+ * etc.
+ *
+ * One example of such classes is {@link OpenShiftRecorderHandler} which sets (and resets) filters or records OCP
+ * state when handling different JUnit events.
+ *
+ * State recording is about downloading several resources logs from OCP.
+ *
+ * Resources are filtered by name provided via {@link OpenShiftRecorder} annotation. Names are turned into regexes
+ * by adding {@code .*} as a suffix. If no name is provided, resources in namespaces (BM and master) are filtered
+ * automatically by recording what resources are seen before test and so on.
+ * <p>
+ * Recorded resources:
+ * <ul>
+ * <li>pods states</li>
+ * <li>deployment configs states</li>
+ * <li>builds states</li>
+ * <li>build configs states</li>
+ * <li>image streams states</li>
+ * <li>stateful sets states</li>
+ * <li>routes states</li>
+ * <li>services states</li>
+ * <li>secrets states</li>
+ * <li>logs of pods</li>
+ * <li>logs of builds</li>
+ * <li>events</li>
+ * </ul>
+ * <p>
+ * OpenShift state is recorded when a test throws an exception. If {@link JUnitConfig#recordAlways()} is true, state is
```

Comment:
This is no longer true.

Reply:
Yes, agree, this should be removed altogether, since it's not relevant for this service class.

Reply:
Fixed.